### PR TITLE
Feature LLMCachingMiddleware

### PR DESCRIPTION
**Pull request by Winston-503**
- Implement `LLMCachingMiddleware` with `ttl` and `cache_limit_size` to cache LLM responses based on request (messages and parameters) and LLM config
- Update docs, add unit and integration tests

### CODE REVIEW

File: council/llm/llm_middleware.py [hunk 1]
```diff
@@ -186,3 +189,102 @@ def __call__(self, llm: LLMBase, execute: ExecuteLLMRequest, request: LLMRequest
...
+
+        return response
+
+    @staticmethod
+    def get_request_hash(request: LLMRequest) -> str:
+        """Convert the request to a hash with hashlib.sha256."""
+        serialized = json.dumps(
+            {
+                # TODO: request.context is not serialized
+                "messages": [str(m) for m in request.messages],  # TODO: serialize messages properly
```
Comment (aflament) [inline council/llm/llm_middleware.py:270]: @Winston-503 do we want:
- to lower case
- suppress spaces
- any other optimization?

Comment (Winston-503) [reply council/llm/llm_middleware.py:270]: I think lowercasing and removing whitespace characters is enough, but we should do it for `content` and also each `data` element

File: tests/integration/llm/test_llm_caching_middleware.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+        return response
+
+    @staticmethod
+    def assert_if_cached(response):
+        assert all(not consumption.unit.startswith("cached_") for consumption in response.result.consumptions)
+
+    @staticmethod
+    def assert_if_not_cached(response):
+        assert response.duration == 0
```
Comment (aflament) [inline tests/integration/llm/test_llm_caching_middleware.py:44]: duration should be > 0 ?

File: tests/integration/llm/test_llm_caching_middleware.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        print(f"\tResponse duration: {response.duration:.3f}s")
+
+        return response
+
+    @staticmethod
+    def assert_if_cached(response):
+        assert all(not consumption.unit.startswith("cached_") for consumption in response.result.consumptions)
+
+    @staticmethod
+    def assert_if_not_cached(response):
```
Comment (aflament) [inline tests/integration/llm/test_llm_caching_middleware.py:43]: I guess I am a bit confused by the name. I would have expect `assert_cached`

Comment (aflament) [reply tests/integration/llm/test_llm_caching_middleware.py:43]: and the opposite for the previous one

File: tests/integration/llm/test_llm_caching_middleware.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    def test_ttl(self):
+        llm_func = self.get_llm_func(ttl=2, cache_limit_size=10)
+
+        response = self.execute_llm_func(llm_func, self.m1, to_print="message_v1")
+        self.assert_if_cached(response)
+
+        response = self.execute_llm_func(llm_func, self.m1, to_print="message_v1 (cached)")
+        self.assert_if_not_cached(response)
+
+        time.sleep(2)
```
Comment (aflament) [inline tests/integration/llm/test_llm_caching_middleware.py:75]: maybe 2+ to avoid flakiness

**Review by aflament — APPROVED** [review]

File: tests/integration/llm/test_llm_caching_middleware.py (hunk 3 above)
Comment (Winston-503) [reply tests/integration/llm/test_llm_caching_middleware.py:43]: True, renamed to `raise_if_cached` / `raise_if_not_cached`

**Review by aflament — APPROVED** [review]
👍

**Pull request merged**